Sources/AsyncAlgorithms/AsyncMerge2Sequence.swift

```diff
@@ -9,201 +9,1069 @@
+                // We have outstanding demand so request the next element
+                return .resumeContinuation(upstreamContinuation: continuation)
+
+            case .merging(let task, let buffer, var upstreamContinuations, let upstreamsFinished, .none):
+                // There is no outstanding demand from the downstream
+                // so we are storing the continuation and resume it once there is demand.
+                state = .modifying
+
+                upstreamContinuations.append(continuation)
+
+                state = .merging(
+                    task: task,
+                    buffer: buffer,
+                    upstreamContinuations: upstreamContinuations,
+                    upstreamsFinished: upstreamsFinished,
+                    downstreamContinuation: nil
+                )
+
+                return .none
+
+            case .upstreamThrew:
+                // Another upstream already threw so we just need to throw from this continuation
+                // which will end the consumption of the upstream.
+
+                return .resumeContinuationWithError(
+                    upstreamContinuation: continuation,
+                    error: CancellationError()
+                )
+
+            case .finished:
+                // Since cancellation is cooperative it might be that child tasks are still getting
+                // suspended even though we already cancelled them. We must tolerate this and just resume
+                // the continuation with an error.
+                return .resumeContinuationWithError(
+                    upstreamContinuation: continuation,
+                    error: CancellationError()
+                )
+
+            case .modifying:
+                preconditionFailure("Invalid state")
+            }
+        }
+
+        /// Actions returned by `elementProduced()`.
+        enum ElementProducedAction {
+            /// Indicates that the downstream continuation should be resumed with the element.
+            case resumeContinuation(
+                downstreamContinuation: CheckedContinuation<Element?, Error>,
+                element: Element
+            )
+            /// Indicates that nothing should be done.
+            case none
+        }
+
+        mutating func elementProduced(_ element: Element) -> ElementProducedAction {
+            switch state {
+            case .initial:
+                // Child tasks that are producing elements are only created after we transitioned to `merging`
+                preconditionFailure("Invalid state")
+
+            case let .merging(task, buffer, upstreamContinuations, upstreamsFinished, .some(downstreamContinuation)):
+                // We produced an element and have an outstanding downstream continuation
+                // this means we can go right ahead and resume the continuation with that element
+                precondition(buffer.isEmpty, "We are holding a continuation so the buffer must be empty")
```

Comment:
this feels like it somehow could not be fully true... do we have tests that validate this?

Reply:
Various tests are validating that indirectly. For this precondition to fail we would need to get into a state where we have a suspended call to `next()` and things in the buffer. However, when we call `next` and have things in the buffer we are returning them right away. We are only suspending if the buffer is empty. Furthermore, since we are holding the lock across the `withUnsafeContinuation` call in `next` nothing can get in-between us checking if the buffer is empty and us suspending.

When any of the upstreams produces a new element, it will acquire the lock and then we come into this case. If we have a continuation we are guaranteed to have an empty buffer since we would have never suspended in the first place if we had something in the buffer.